Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -2250,17 +2250,25 @@ void CGContextClearRect(CGContextRef context, CGRect rect) {
     return S_OK;
 }
 
-HRESULT __CGContext::DrawGeometry(_CGCoordinateMode coordinateMode, ID2D1Geometry* pGeometry, CGPathDrawingMode drawMode) {
+HRESULT __CGContext::DrawGeometry(_CGCoordinateMode coordinateMode,
+                                  ID2D1Geometry* pGeometry,
+                                  CGPathDrawingMode drawMode,
+                                  bool isPathGeometry) {
     ComPtr<ID2D1Geometry> geometry(pGeometry);
-    return Draw(coordinateMode, nullptr, [geometry, drawMode](CGContextRef context, ID2D1DeviceContext* deviceContext) {
+    return Draw(coordinateMode, nullptr, [geometry, drawMode, isPathGeometry](CGContextRef context, ID2D1DeviceContext* deviceContext) {
         auto& state = context->CurrentGState();
         if (drawMode & kCGPathFill) {
             state.fillBrush->SetOpacity(state.alpha);
 
             ComPtr<ID2D1Geometry> geometryToFill;
-            D2D1_FILL_MODE d2dFillMode = (drawMode & kCGPathEOFill) == kCGPathEOFill ? D2D1_FILL_MODE_ALTERNATE : D2D1_FILL_MODE_WINDING;
-            RETURN_IF_FAILED(_CGConvertD2DGeometryToFillMode(geometry.Get(), d2dFillMode, &geometryToFill));
-
+            // If the geometry is a path and has requested the default fill mode we can avoid the fill mode conversion step.
+            if (!isPathGeometry || drawMode == kCGPathEOFill) {
```

Comment:
I'm not sure what you mean by that. drawMode here is passed in. We already know we have to fill the geometry but if it's a winding fill and a path geometry we're allowed to skip this conversion step since CGPath has already set the fill mode by default.